File: app/src/main/java/com/github/ilikeyourhat/kudoku/solving/deduction/algorithm/NakedSinglesAlgorithm.java
package com.github.ilikeyourhat.kudoku.solving.deduction.algorithm;

import com.github.ilikeyourhat.kudoku.model.Field;
import com.github.ilikeyourhat.kudoku.model.Region;
import com.github.ilikeyourhat.kudoku.model.hint.SudokuHintGrid;
import com.github.ilikeyourhat.kudoku.model.Field;
import com.github.ilikeyourhat.kudoku.model.Region;
import com.github.ilikeyourhat.kudoku.model.hint.SudokuHintGrid;

import java.util.List;
import java.util.Set;

public class NakedSinglesAlgorithm extends DeductionAlgorithm {

    public static class Factory implements DeductionAlgorithm.Factory {

        public NakedSinglesAlgorithm instance(List<Region> regions, SudokuHintGrid sudokuHintGrid) {
            return new NakedSinglesAlgorithm(regions, sudokuHintGrid);
        }

    }

    private NakedSinglesAlgorithm(List<Region> regions, SudokuHintGrid possibilities) {
        super(regions, possibilities);
    }

    @Override
    protected void solve(Region region) {
        for (Field field : region.emptyFields()) {
            Set<Integer> values = possibilities.forField(field);
            if (values.size() == 1) {
                int singleValue = (int) values.toArray()[0];
                field.set(singleValue);
                notifyChange();
            }
        }
    }

}

File: app/src/main/java/com/github/ilikeyourhat/kudoku/solving/deduction/algorithm/NakedSinglesAlgorithm.kt
package com.github.ilikeyourhat.kudoku.solving.deduction.algorithm

import com.github.ilikeyourhat.kudoku.model.Region
import com.github.ilikeyourhat.kudoku.model.hint.SudokuHintGrid

class NakedSinglesAlgorithm (
    regions: List<Region>,
    possibilities: SudokuHintGrid
) : DeductionAlgorithm(regions, possibilities) {

    class Factory : DeductionAlgorithm.Factory {
        override fun instance(regions: List<Region>, possibilities: SudokuHintGrid): NakedSinglesAlgorithm {
            return NakedSinglesAlgorithm(regions, possibilities)
        }
    }

    override fun solve(region: Region): Boolean {
        for (field in region.emptyFields()) {
            val values = possibilities.forField(field)
            if (values.size == 1) {
                field.set(values.single())
                return true
            }
        }
        return false
    }
}